lib/spack/spack/cmd/log.py

```diff
@@ -0,0 +1,103 @@
+# Copyright 2013-2024 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+import os
+import shutil
+import tempfile
+import time
+
+import llnl.util.filesystem as fs
+import llnl.util.tty as tty
+
+import spack.cmd
+import spack.environment as ev
+import spack.util.compression as compression
+from spack.cmd.common import arguments
+
+description = "print out logs for packages"
+section = "basic"
+level = "long"
+
+
+def setup_parser(subparser):
+    subparser.add_argument(
+        "--continuous", action="store_true", help="for stage logs, poll for new content"
+    )
+
+    arguments.add_common_arguments(subparser, ["spec"])
+
+
+def dump_build_log(path, continuous):
+    with open(path, "r") as fstream:
+        line = fstream.readline()
+        while True:
+            if line:
+                print(line.strip())
+            elif continuous:
+                time.sleep(0.2)
+            else:
+                break
+            line = fstream.readline()
+
+
+def log(parser, args):
+    specs = spack.cmd.parse_specs(args.spec)
+
+    if not specs:
+        tty.die("You must supply a spec.")
+
+    if len(specs) != 1:
+        tty.die("Too many specs.  Supply only one.")
+
+    env = ev.active_environment()
+
+    if env:
+        spec = spack.cmd.matching_spec_from_env(specs[0])
+    else:
+        spec = specs[0].concretized()
+
+    if spec.installed:
+        log_path = spec.package.install_log_path
+    elif os.path.exists(spec.package.stage.path):
+        dump_build_log(spec.package.log_path, args.continuous)
+        return
+    else:
+        tty.die(f"{specs[0]} is not installed or staged")
+
+    if not os.path.exists(log_path):
+        tty.die(f"No logs are available for {specs[0]}")
+
+    compression_ext = compression.extension_from_file(log_path)
+    fstream = None
+    temp_dir = None
+    try:
+        if not compression_ext:
+            fstream = open(log_path, "r")
+        else:
+            decompressor = compression.decompressor_for(log_path, extension=compression_ext)
```

Comment:
Just use `gzip.open`

Reply:
I like the part about opening `stdout` in binary mode (no need for decoding)

I disagree about assuming this is a `.gz` - that seems overly-tied to decisions made outside of this. The current implementation is straightforward as is ("give me the decompressor for the log file" is easier than "this file is obviously a .gz because of code written elsewhere and I'm going to run with that") and also handles the case of older (not-compressed) log files.

Reply:
If you care so much about that then just branch on the extension: .gz for compressed logs, .txt/.log for uncompressed logs, else error. If somebody would ever add support for .bzip / .zstd / whatever, which is never going to happen anyways, your test will error.

I _really_ dislike how you're writing this. I just absolutely don't want to see a "solution" where you dump a file to stdout that does a double pass over the data and involves a temporary directory and a temporary file.

If you really like to implement it that way then I'll let @tgamblin review it.

Reply:
> If you care so much about future decompression schemes

I don't, and to be clear I explicitly don't care about the *current* decompression scheme, which is why I thought it was nice to shell out to an existing bit of code which handles that for us (the `chdir` stuff is mildly clunky, but completely straightforward, and could have been shunted into `filesystem`). 

> that does a double pass over the data, involves creating a temporary directory and a temporary file, as well as chdir twice. If you don't have to store decompressed data, then don't do it, 

None of these things seem particularly problematic for a command that is explicitly designed to dump everything into the user's shell, which is why I think some pushback was warranted. Since you insist, and I see no problems with doing this, it is now done.